### PR TITLE
Remove callbacks from Rename api.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Rename has two callbacks that can be used by internal clients when invoking the API.  The first allows the client to filter the list of locations to update to prevent edits in places that are undesirable for that feature.  The second allows the feature to opt out of some forms of conflict detection.  This is useful when the feature knows what it is doing will cause temporary conflicts and it doesn't want rename's conflict fixer to fix up those locations.

These callbacks are problematic from an OOP perspective.  We don't want the oop server calling back into the host to compute this information.  

To fix this, i've adjusted things to be more OOP friendly by allowing pure data to be passed in instead of a callback.  The pure data is less functional than before. However, as we only have one single feature using this functionality, we can cater to its need, instead of a more flexible approach that isn't oopable.

### CODE REVIEW

File: src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs [hunk 1]
```diff
@@ -244,27 +245,35 @@ private async Task<Result> EncapsulateFieldAsync(IFieldSymbol field, Document do
...
+                        location => IntersectsWithAny(location, constructorLocations));
+
+                    solution = await Renamer.RenameAsync(
+                        insideLocations, finalFieldName, cancellationToken: cancellationToken).ConfigureAwait(false);
                     document = solution.GetDocument(document.Id);
 
                     var compilation = await document.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
 
                     field = field.GetSymbolKey().Resolve(compilation, cancellationToken: cancellationToken).Symbol as IFieldSymbol;
+                    constructorLocations = GetConstructorLocations(field.ContainingType);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs:264]: this was a bug i noticed.  after performing the first rename, we weren't recomputing the spans of the constructors in the new solution.

File: src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs [hunk 2]
```diff
@@ -244,27 +245,35 @@ private async Task<Result> EncapsulateFieldAsync(IFieldSymbol field, Document do
...
-                        location => constructorSyntaxes.Any(c => c.Span.IntersectsWith(location.SourceSpan)),
-                        cancellationToken: cancellationToken).ConfigureAwait(false);
+                    var initialLocations = await Renamer.GetRenameLocationsAsync(
+                        solution, SymbolAndProjectId.Create(field, projectId), solution.Options, cancellationToken).ConfigureAwait(false);
+
+                    var insideLocations = initialLocations.Filter(
+                        location => IntersectsWithAny(location, constructorLocations));
+
+                    solution = await Renamer.RenameAsync(
+                        insideLocations, finalFieldName, cancellationToken: cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs:258]: the overload that has the rename engine filter is gone.  now, features can jut ask for the locations that will be renamed, filter them themselves, then ask to rename the filtered set.  

Both GetRenameLocationsAsync and RenameAsync will be OOPable in an upcoming PR.

File: src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs [hunk 3]
```diff
@@ -120,10 +120,20 @@ private async Task<Solution> ProcessResultAsync(CodeFixContext context, Diagnost
...
+            // To address this, we let rename know that there is no conflict if the new symbol it resolves to is the
+            // same as the property we're trying to get the references pointing to.
+
+            var updatedSolution = await Renamer.RenameAsync(
+                fieldLocations.Filter(
+                    location => !location.IntersectsWith(declaratorLocation) &&
+                                CanEditDocument(solution, location.SourceTree, linkedFiles, canEdit)),
+                propertySymbol.Name,
+                nonConflictSymbols: ImmutableHashSet.Create<ISymbol>(propertySymbol),
                 cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs:137]: two changes here.  first, we have already gotten the rename locations above, so here we're only filtering it.  second, instead of passing a callback function to not error on symbols that bind to the property post-rename, we just pass the property into the api, and the api will do that check.

File: src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs [hunk 4]
```diff
@@ -324,33 +334,6 @@ private async Task<SyntaxNode> FormatAsync(SyntaxNode newRoot, Document document
...
-                    {
-                        // No conflict.
-                        return false;
-                    }
-                }
-            }
-
-            // Just do the default check.
-            return null;
-        }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs:352]: this moves into conflict detection.

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs [hunk 5]
```diff
@@ -374,13 +380,12 @@ private async Task DebugVerifyNoErrorsAsync(ConflictResolution conflictResolutio
...
-                            }
+                                hasConflict =
+                                    !IsConflictFreeChange(newReferencedSymbols, nonConflictSymbols) &&
+                                    await CheckForConflictAsync(conflictResolution, renamedSymbolInNewSolution, conflictAnnotation, newReferencedSymbols).ConfigureAwait(false);
 
-                            if (!hasConflict && !conflictAnnotation.IsInvocationExpression)
-                            {
-                                hasConflict = LocalVariableConflictPerLanguage((SyntaxToken)tokenOrNode, newDocument, newReferencedSymbols);
+                                if (!hasConflict && !conflictAnnotation.IsInvocationExpression)
+                                    hasConflict = LocalVariableConflictPerLanguage((SyntaxToken)tokenOrNode, newDocument, newReferencedSymbols);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs:388]: this moved into the above block because it references newReferencedSymbols  (which is only assigned in this block).

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs [hunk 6]
```diff
@@ -611,31 +642,25 @@ private async Task DebugVerifyNoErrorsAsync(ConflictResolution conflictResolutio
...
                 if (conflictAnnotation.IsInvocationExpression)
                 {
-                    IEnumerable<ISymbol> invocationReferencedSymbols = null;
                     if (tokenOrNode.IsNode)
                     {
-                        invocationReferencedSymbols = SymbolsForEnclosingInvocationExpressionWorker((SyntaxNode)tokenOrNode, newDocumentSemanticModel, _cancellationToken);
-                    }
-
-                    if (invocationReferencedSymbols != null)
-                    {
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs:627]: these blocks merged.

File: src/Workspaces/Core/Portable/Rename/RenameLocations.cs [hunk 7]
```diff
@@ -218,5 +218,12 @@ internal async Task<RenameLocations> FindWithUpdatedOptionsAsync(OptionSet optio
 
             return new SearchResult(locations, implicitLocations, referencedSymbols);
         }
+
+        public RenameLocations Filter(Func<Location, bool> filter)
+            => new RenameLocations(
+                this.Locations.Where(loc => filter(loc.Location)).ToSet(),
+                this.SymbolAndProjectId, this.Solution,
+                this.ReferencedSymbols, this.ImplicitLocations.Where(loc => filter(loc.Location)),
+                this.Options);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Rename/RenameLocations.cs:227]: this moved out of Renamer below.

File: src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs (hunk 2 above)
Comment (mavasani) [reply src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs:258]: Nit: consider extracting out the common code into a shared helper parameterized by `locationFilter`, say

```csharp
                  static async Task<Solution> RenameSymbolAsync(locationFilter)
                  {
                    var initialLocations = await Renamer.GetRenameLocationsAsync(
                        solution, SymbolAndProjectId.Create(field, projectId), solution.Options, cancellationToken).ConfigureAwait(false);

                    var insideLocations = initialLocations.Filter(locationFilter);

                    return await Renamer.RenameAsync(
                        insideLocations, finalFieldName, cancellationToken: cancellationToken).ConfigureAwait(false);
                  }

File: src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs [hunk 8]
```diff
@@ -275,6 +284,20 @@ private async Task<Result> EncapsulateFieldAsync(IFieldSymbol field, Document do
             }
         }
 
+        private bool IntersectsWithAny(Location location, ISet<Location> constructorLocations)
```
Comment (mavasani) [inline src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs:287]:
```suggestion
        private static bool IntersectsWithAny(Location location, ISet<Location> constructorLocations)
```

File: src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs [hunk 9]
```diff
@@ -120,10 +120,20 @@ private async Task<Solution> ProcessResultAsync(CodeFixContext context, Diagnost
...
-                symbols => HasConflict(symbols, propertySymbol, compilation, cancellationToken),
+            //
+            // We're asking the rename API to update a bunch of references to an existing field to the same name as an
+            // existing property.  Rename will often flag this situation as an unresolvable conflict because the new
+            // name won't bind to the field anymore.
+            //
+            // To address this, we let rename know that there is no conflict if the new symbol it resolves to is the
+            // same as the property we're trying to get the references pointing to.
+
+            var updatedSolution = await Renamer.RenameAsync(
```
Comment (mavasani) [inline src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs:131]: FYI: I actually attempted to port this analyzer/code fix to shared layer, but it used an internal Rename API, so it did not get to complete it. It would be good that all APIs used here are public APIs, or at least we have a plan to make them public.

File: src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs [hunk 10]
```diff
@@ -115,7 +115,7 @@ public async Task AddDiagnosticWithMappedFileToManyDocumentsTestAsync()
             Assert.Contains(results[1].Diagnostics, d => d.Code == "doc2Diagnostic");
         }
 
-        [Fact]
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/43046")]
```
Comment (mavasani) [inline src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:118]: Is this an unrelated change?

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:118]: yes.  disabling flakey test.

File: src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs:131]: yes. i will file issue on that.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs:131]: https://github.com/dotnet/roslyn/issues/43055

File: src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs:258]: will do in next PR.  thanks!

File: src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/EncapsulateField/AbstractEncapsulateFieldService.cs:287]: will do in next PR.  thanks!
